### ISSUE TITLE
WPCOM API: use Jetpack to authenticate when querying WordPress.com REST API endpoints

### ISSUE THREAD

**Issue by TravisRothlisberger** [Automattic/jetpack#8347]
#### Steps to reproduce the issue
This originated as a [support request on the Jetpack plugin support forum](https://wordpress.org/support/topic/development-using-jetpack-authentication-to-access-the-wordpress-com-api/#post-9774944) but I was redirected here. See that discussion for additional details.

I am trying to use the authentication framework in Jetpack to reach the [`read/following/mine`](https://developer.wordpress.com/docs/api/1.1/get/read/following/mine/) and [`read/liked`](https://developer.wordpress.com/docs/api/1.1/get/read/liked/) endpoints on the Wordpress.com REST API, with the end goal of producing a plugin that does something like the '[Blogs I Follow](https://en.support.wordpress.com/widgets/blogs-i-follow-widget/)' and '[Posts I Like](https://en.support.wordpress.com/widgets/posts-i-like-widget/)' widgets, but for a self-hosted site.

I started by following the advice to call `Jetpack_Client::wpcom_json_api_request_as_blog('read/following/mine')` and `Jetpack_Client::wpcom_json_api_request_as_blog('read/liked')` from a simple plugin and received a 403 response of "That API call is not allowed for this account". Upon review of the code I found that this call was using a [`blog_token`](https://github.com/Automattic/jetpack/blob/master/class.jetpack-data.php#L31) and theorized that this could account for the authentication failure.

So I tweaked `Jetpack_Client::wpcom_json_api_request_as_blog` to allow me to pass a `user_id` argument through and supplied the value from both `Jetpack_Options::get_option('master_user')` and `get_current_user_id()` (which are the same) as the `user_id`. This produced a successful response from the `read/following/mine` endpoint but returned an empty subscriptions array despite my having a large collection of followed blogs accessed in the Wordpress Reader. The `read/liked` endpoint returned a different 403 response of "An active access token must be used to query information about the current user."

#### What I expected
I was hoping that the authentication managed by Jetpack would give me access to the relevant endpoints for the Wordpress.com API without having to built up a separate framework for managing authentication for a plugin, which appears to be a relatively substantial task without clear precedent outside of (possibly) Jetpack. I'm hopeful that perhaps this is still true, but I don't know where to look next.

#### What happened instead
See responses noted in the steps above.

Please advise on the feasibility of reaching the desired endpoints via the Jetpack authentication framework and on the proper modes of utilizing the framework to successfully access those endpoints. Thanks.

**Comment by jeherve** [comment]
> I was hoping that the authentication managed by Jetpack would give me access to the relevant endpoints for the Wordpress.com API without having to built up a separate framework for managing authentication for a plugin

We currently limit the list of endpoints that can be accessed when authenticating via Jetpack, since we do not query all endpoints in Jetpack. Which endpoints would you need to be able to query? Just `read/following/mine` and `read/liked`?

**Comment by TravisRothlisberger** [comment]
> Which endpoints would you need to be able to query? Just read/following/mine and read/liked?
Yes, those are the only endpoints I foresee wanting to query - assuming I am correctly interpreting those as corresponding to the content one would find in the 'Blogs I Follow' and 'Posts I Like' widgets on wordpress.com.

**Comment by jeherve** [comment]
Thank you.

This is being addressed internally in D8789-code.

**Comment by TravisRothlisberger** [comment]
Thanks. I don't know what "D8789-code" is referencing, so should I just assume that the only visibility comes through status updates on this issue?

**Comment by jeherve** [comment]
Yes, I will let you know as soon as we make the changes upstream!

**Comment by TravisRothlisberger** [comment]
@brbrr Thank you for removing the stale label. Do you by chance have any insight into whether this (or the corresponding D8789 code) will be implemented any time soon? I tried again using the new wpcom_json_api_request_as_user function and received a response recognizing my user account, but received {"code":"rest_no_route","message":"No route was found matching the URL and request method","data":{"status":404}} for both /read/liked and /read/following/mine. Thanks.

**Comment by brbrr** [comment]
Hi there,
Unfortunately, we are not planning to address this issue in near future. We will update this ticket as soon as we going to work on this.

**Comment by jeffpaul** [comment]
Similar to @TravisRothlisberger interests, I'd like to see this exposed.  @jeherve I'm guessing still stalled inside a8c on this (though perhaps if prior stalls was lack of engineering perhaps now with some AI guidance is more feasible)?

**Comment by jeherve** [comment]
Sorry for the lack of update on this one. It slipped through the cracks :) Thank you for bringing it back to my notifcations!

We ended up not opening the v1.1 `read/liked` and `read/following/mine` endpoints to Jetpack's site token. They return data for a WordPress.com user, and a site token isn't useful there because it can only be used to return data for a site.

That said, both of the endpoints mentioned above are now available on the v2 API, so one can make a user-authenticated request to those endpoints like so:

```php
use Automattic\Jetpack\Connection\Client;

$following = Client::wpcom_json_api_request_as_user( '/me/following', '2' );
$liked     = Client::wpcom_json_api_request_as_user( '/liked-posts', '2' );
```

This is what Jetpack's Blogroll block uses today for example:
https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/extensions/blocks/blogroll/use-subscriptions.js#L24-L25

I think I can close this issue now, this should be just what you needed. Let me know if you have any other questions!
